pkg/kubelet/cm/node_allocatable.go

```diff
@@ -0,0 +1,216 @@
+	NodeAllocatableEnforcementKey    = "pods"
+	SystemReservedEnforcementKey     = "system-reserved"
+	KubeReservedEnforcementKey       = "kube-reserved"
+	// Constants for identifying various types of Node Allocatable reservations.
+	forScheduling = iota // 0
+	forEnforcement
+)
+
+func (cm *containerManagerImpl) createNodeAllocatableCgroups() error {
+	cgroupConfig := &CgroupConfig{
+		Name: CgroupName(cm.cgroupRoot),
+	}
+	if cm.cgroupManager.Exists(cgroupConfig.Name) {
+		return nil
+	}
+	if err := cm.cgroupManager.Create(cgroupConfig); err != nil {
+		glog.Errorf("Failed to create %q cgroup", cm.cgroupRoot)
+		return err
+	}
+	return nil
+}
+
+// Enforce Node Allocatable Cgroup settings.
+func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
+	nc := cm.NodeConfig.NodeAllocatableConfig
+	nodeAllocatable := cm.getNodeAllocatableInternal(forEnforcement)
+
+	glog.V(4).Infof("Attempting to enforce Node Allocatable with config: %+v", nc)
+	glog.V(4).Infof("Node Allocatable resources: %+v", nodeAllocatable)
+	// Create top level cgroups for all pods if necessary.
+	if nc.EnforceNodeAllocatable.Has(NodeAllocatableEnforcementKey) {
+		cgroupConfig := &CgroupConfig{
+			Name:               CgroupName(defaultNodeAllocatableCgroupName),
+			ResourceParameters: getCgroupConfig(nodeAllocatable),
+		}
+		// If Node Allocatable is enforced on a node that has not been drained or is updated on an existing node to a lower value,
+		// existing memory usage across pods might be higher that current Node Allocatable Memory Limits.
+		// Pod Evictions are expected to bring down memory usage to below Node Allocatable limits.
+		// Until evictions happen retry cgroup updates.
+		go func(node *v1.Node, r record.EventRecorder) {
+			for {
+				glog.V(4).Infof("Attempting to update Node Allocatable cgroup with %d cpu shares and %d bytes of memory", cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory)
+				err := cm.cgroupManager.Update(cgroupConfig)
+				if err == nil {
+					r.Event(node, v1.EventTypeNormal, events.SuccessfulNodeAllocatableEnforcement, "Updated Node Allocatable limit across pods")
+				}
+				message := fmt.Sprintf("Failed to update limits on cgroup %q: %v", err)
+				r.Event(node, v1.EventTypeWarning, events.FailedNodeAllocatableEnforcement, message)
+				glog.Error(message)
+				// TODO: Set Limits to be a few bytes greater than current memory usage to try to force kernel to reclaim memory.
+				// Retry after a minute.
+				time.Sleep(time.Minute)
+			}
+		}(cm.nodeInfo, cm.recorder)
+	}
+	// Now apply kube reserved and system reserved limits if required.
+	if nc.EnforceNodeAllocatable.Has(SystemReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing system reserved on cgroup %q with limits: %+v", nc.SystemReservedCgroupName, nc.SystemReserved)
+		if err := enforceExistingCgroup(cm.cgroupManager, nc.SystemReservedCgroupName, nc.SystemReserved); err != nil {
```

Comment:
Actually, I think I recall why a retry isn't useful. System reserved and kube reserved cgroups are not managed by kubelet and so it cannot reclaim memory in the case of failures. So it doesn't make sense to retry. Adding an event makes sense though, but kubelet initialization would fail to make the error obvious.